File: research/scripts/uab_census.py
"""Certificate census for general (a,b): do the four Theorem-H layers close
higher exponent grids, and how do the residue patterns stabilize?

For each grid (a,b): enumerate classes (j,k,sg), leaves, grading kill.
For survivors, extract the PATTERN of each relation's minimal p-layer:
the multiset of (k, sg, coeff) at the tying j — shifted so min k = 0 —
plus the same on the q-side. Patterns, not leaves, feed the certificate
computation, so the count stays small at every grid.
"""
import itertools, pickle, sys
from collections import Counter

def classes(a, b):
    out = []
    for j in range(a+1):
        for k in range(b+1):
            if (j, k) == (0, 0): continue
            sgs = (1,) if (j == 0 or k == 0) else (1, -1)
            for sg in sgs:
                out.append((j, k, sg))
    return out

def run(a, b):
    els = classes(a, b)
    n = len(els)
    pexp = {e: 2*(a - e[0]) for e in els}
    qexp = {e: 2*(b - e[1]) for e in els}
    def lone(rel):
        for exp in (pexp, qexp):
            vs = [exp[e] for e, _ in rel]
            m = min(vs)
            if vs.count(m) == 1: return True
        return False
    total = killed = 0
    pat1 = Counter()   # (p-side pattern of R1, p-side pattern of R2)
    surv = 0
    for combo in itertools.combinations(range(n), 4):
        for perm in itertools.permutations(combo):
            ea, eb, ec, ed = [els[i] for i in perm]
            for e2, e3, e4 in itertools.product((1,-1), repeat=3):
                total += 1
                r1 = [(ec, e3), (ed, e4), (ea, -2)]
                r2 = [(ec, e3), (ed, -e4), (eb, -2*e2)]
                if lone(r1) or lone(r2):
                    killed += 1
                    continue
                surv += 1
                def ppat(rel):
                    m = min(pexp[e] for e, _ in rel)
                    lay = sorted((e[1], e[2], g) for e, g in rel if pexp[e] == m)
                    k0 = min(k for k, _, _ in lay)
                    return tuple((k - k0, sg, g) for k, sg, g in lay)
                pat1[(ppat(r1), ppat(r2))] += 1
    return total, killed, surv, pat1

grids = [(2,2), (3,2), (2,3), (3,3)]
allpats = {}
for (a, b) in grids:
    total, killed, surv, pats = run(a, b)
    allpats[(a,b)] = pats
    print(f"({a},{b}): {total} leaves, grading-killed {killed} ({100*killed/total:.1f}%), "
          f"survivors {surv}, distinct p-layer pattern pairs: {len(pats)}")
# stabilization: which patterns at (3,3) are new vs (2,2)?
base = set(allpats[(2,2)])
for g in grids[1:]:
    new = set(allpats[g]) - base
    print(f"pattern pairs in {g} not present in (2,2): {len(new)} / {len(allpats[g])}")
pickle.dump(allpats, open('uab_patterns.pkl', 'wb'))

File: research/scripts/rs/src/bin/uab_census.rs
// Pattern census for general (a,b): confirm the minimal-layer pattern
// universe saturates. A pattern pair is the (shifted) multiset of
// (k-offset, sign, coeff) in each relation's minimal p-layer.
use std::collections::HashSet;

type Class = (i32, i32, i32); // (j, k, sg)
type Pat = Vec<(i32, i32, i64)>;

fn classes(a: i32, b: i32) -> Vec<Class> {
    let mut out = Vec::new();
    for j in 0..=a {
        for k in 0..=b {
            if j == 0 && k == 0 { continue; }
            if j == 0 || k == 0 { out.push((j, k, 1)); }
            else { out.push((j, k, 1)); out.push((j, k, -1)); }
        }
    }
    out
}

fn ppat(rel: &[(Class, i64)], a: i32) -> Pat {
    let m = rel.iter().map(|&((j, _, _), _)| 2 * (a - j)).min().unwrap();
    let mut lay: Vec<(i32, i32, i64)> = rel.iter()
        .filter(|&&((j, _, _), _)| 2 * (a - j) == m)
        .map(|&((_, k, sg), g)| (k, sg, g)).collect();
    let k0 = lay.iter().map(|&(k, _, _)| k).min().unwrap();
    for t in lay.iter_mut() { t.0 -= k0; }
    lay.sort();
    lay
}

fn run(a: i32, b: i32) -> (u64, u64, HashSet<(Pat, Pat)>) {
    let els = classes(a, b);
    let n = els.len();
    let mut total = 0u64;
    let mut surv = 0u64;
    let mut pats = HashSet::new();
    let lone = |rel: &[(Class, i64)]| {
        for f in [true, false] {
            let vs: Vec<i32> = rel.iter().map(|&((j, k, _), _)|
                if f { 2 * (a - j) } else { 2 * (b - k) }).collect();
            let m = *vs.iter().min().unwrap();
            if vs.iter().filter(|&&v| v == m).count() == 1 { return true; }
        }
        false
    };
    for ia in 0..n { for ib in 0..n { for ic in 0..n { for id in 0..n {
        if ia == ib || ia == ic || ia == id || ib == ic || ib == id || ic == id { continue; }
        let (ea, eb, ec, ed) = (els[ia], els[ib], els[ic], els[id]);
        for e2 in [1i64, -1] { for e3 in [1i64, -1] { for e4 in [1i64, -1] {
            total += 1;
            let r1 = [(ec, e3), (ed, e4), (ea, -2)];
            let r2 = [(ec, e3), (ed, -e4), (eb, -2 * e2)];
            if lone(&r1) || lone(&r2) { continue; }
            surv += 1;
            pats.insert((ppat(&r1, a), ppat(&r2, a)));
        }}}
    }}}}
    (total, surv, pats)
}

fn main() {
    let grids = [(2, 2), (2, 3), (3, 3), (4, 3), (3, 4), (4, 4), (5, 3), (2, 5), (5, 5)];
    let mut base: Option<HashSet<(Pat, Pat)>> = None;
    for (a, b) in grids {
        let (total, surv, pats) = run(a, b);
        let newc = match &base {
            Some(bs) => pats.difference(bs).count(),
            None => 0,
        };
        println!("({a},{b}): {total} leaves, survivors {surv}, patterns {}, new vs (2,3)-base: {newc}",
                 pats.len());
        if (a, b) == (2, 3) { base = Some(pats); }
    }
}
